test/integration/apiserver/admissionwebhook/match_conditions_test.go

```diff
@@ -0,0 +1,674 @@
+				},
+			},
+			pods: []*corev1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test2",
+						Namespace: "default",
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "test",
+								Image: "test",
+							},
+						},
+					},
+				},
+			},
+			matchedPods: []*corev1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test2",
+						Namespace: "default",
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "test",
+								Image: "test",
+							},
+						},
+					},
+				},
+			},
+			expectError: false,
+		},
+		{
+			name: "invalid field should error",
+			matchConditions: []admissionregistrationv1.MatchCondition{
+				{
+					Name:       "old-object-is-null",
+					Expression: "imnotafield == null",
+				},
+			},
+			pods:        []*corev1.Pod{},
+			matchedPods: []*corev1.Pod{},
+			expectError: true,
+		},
+		{
+			name: "missing expression should error",
+			matchConditions: []admissionregistrationv1.MatchCondition{
+				{
+					Name: "old-object-is-null",
+				},
+			},
+			pods:        []*corev1.Pod{},
+			matchedPods: []*corev1.Pod{},
+			expectError: true,
+		},
+		{
+			name: "missing name should error",
+			matchConditions: []admissionregistrationv1.MatchCondition{
+				{
+					Expression: "imnotafield == null",
+				},
+			},
+			pods:        []*corev1.Pod{},
+			matchedPods: []*corev1.Pod{},
+			expectError: true,
+		},
+		{
+			name: "empty name should error",
+			matchConditions: []admissionregistrationv1.MatchCondition{
+				{
+					Name:       "",
+					Expression: "imnotafield == null",
+				},
+			},
+			pods:        []*corev1.Pod{},
+			matchedPods: []*corev1.Pod{},
+			expectError: true,
+		},
+		{
+			name: "empty expression should error",
+			matchConditions: []admissionregistrationv1.MatchCondition{
+				{
+					Name:       "test-empty-expression",
+					Expression: "",
+				},
+			},
+			pods:        []*corev1.Pod{},
+			matchedPods: []*corev1.Pod{},
+			expectError: true,
+		},
+	}
+
+	for _, testcase := range testcases {
+		t.Run(testcase.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.AdmissionWebhookMatchConditions, true)()
+
+			roots := x509.NewCertPool()
+			if !roots.AppendCertsFromPEM(localhostCert) {
+				t.Fatal("Failed to append Cert from PEM")
+			}
+			cert, err := tls.X509KeyPair(localhostCert, localhostKey)
+			if err != nil {
+				t.Fatalf("Failed to build cert with error: %+v", err)
+			}
+
+			recorder := &admissionRecorder{requests: []*admissionv1.AdmissionRequest{}}
+			upCh := recorder.Reset()
+
+			webhookServer := httptest.NewUnstartedServer(newMatchConditionHandler(recorder))
+			webhookServer.TLS = &tls.Config{
+				RootCAs:      roots,
+				Certificates: []tls.Certificate{cert},
+			}
+			webhookServer.StartTLS()
+			defer webhookServer.Close()
```

Comment:
nit: I think you can reuse a single server across all test cases

Reply:
moved